subsys/bluetooth/audio/mcc.c

```diff
@@ -1195,19 +1182,283 @@ static uint8_t discover_include_func(struct bt_conn *conn,
 }
 #endif /* CONFIG_BT_MCC_OTS */
 
+/* Start discovery of included services */
+static void discover_included(struct bt_conn *conn)
+{
+	int err;
+
+	discover_params.start_handle = cur_mcs_inst->start_handle;
+	discover_params.end_handle = cur_mcs_inst->end_handle;
+	discover_params.type = BT_GATT_DISCOVER_INCLUDE;
+	discover_params.func = discover_include_func;
 
-/* This function is called when characteristics are found.
- * The function will store handles, and optionally subscribe to, GMCS
- * characteristics.
+	BT_DBG("Start discovery of included services");
+	err = bt_gatt_discover(conn, &discover_params);
+	if (err) {
+		discovery_failure(conn, "Discover of included service failed", err);
+	}
+}
+
+static void subscribe_mcs_char_func(struct bt_conn *conn);
+
+static void subscription_cb(struct bt_conn *conn, uint8_t err,
+			    struct bt_gatt_subscribe_params *params)
+{
+	if (err) {
+		discovery_failure(conn, "Subscription callback error", err);
+	}
+
+	BT_DBG("Subscribed: value handle: %d, ccc handle: %d",
+	       params->value_handle, params->ccc_handle);
+
+	/* Subscribe to next characteristic */
+	subscribe_mcs_char_func(conn);
+}
+
+/* Subscribe to a characteristic - helper function */
+static int do_subscribe(struct bt_conn *conn, uint16_t handle,
+			 struct bt_gatt_subscribe_params *sub_params)
+{
+	/* With ccc_handle == 0 it will use auto discovery */
+	sub_params->ccc_handle = 0;
+	sub_params->end_handle = cur_mcs_inst->end_handle;
+	sub_params->value = BT_GATT_CCC_NOTIFY;
+	sub_params->value_handle = handle;
+	sub_params->notify = mcs_notify_handler;
+	sub_params->subscribe = subscription_cb;
+
+	BT_DBG("Subscring to handle %d", handle);
+	return bt_gatt_subscribe(conn, sub_params);
+}
+
+/* This function will subscribe to GMCS CCCDs.
  * After this, the function will start discovery of included services.
  */
+static void subscribe_mcs_char_func(struct bt_conn *conn)
+{
+	struct bt_gatt_subscribe_params *sub_params = NULL;
+	int err = 0;
+
+	/* The characteristics may be in any order on the server, and
+	 * not all of them may exist.
+	 * For each of the characteristics we want to subscribe to
+	 * - check if we have a handle for it
+	 * - check if we have already subscribed to it (using the
+	 * discover_params pointer as a flag)
+	 */
+
+	if (cur_mcs_inst->player_name_handle &&
+	    cur_mcs_inst->player_name_sub_params.disc_params == NULL) {
+		sub_params = &cur_mcs_inst->player_name_sub_params;
+		sub_params->disc_params = &discover_params;
+		err = do_subscribe(conn, cur_mcs_inst->player_name_handle, sub_params);
+		if (err) {
+			discovery_failure(conn, "Could not subscribe", err);
+		}
+		return;
+	}
```

Comment:
I did.
I think I chose to do it this way because of the #ifdef-s that break that pattern - they come in between the else and the following if.

Normally, we have
```
if (<some_condition>) {
        do_something()
} else if (<some_other_condition>) {
        do_something_else()
}
```
But here it would become
```
if (<some_condition>) {
        do_something()
} else 
#ifdef MCC_OTS
if (<some_other_condition>) {
        do_something_else()
}
```

Reply:
Right, that makes sense. The numerous `#ifdef`'s would be a hinder for that. My suggestion from https://github.com/zephyrproject-rtos/zephyr/pull/45762#discussion_r875827186 would fix that though.